### ISSUE TITLE
Clean up action names

### ISSUE THREAD

**Issue by fasterthanlime** [itchio/itch#139]
I'm still a Flux noob pretty much but having inverted sentences like `cave_queue` is confusing, I can never remember which is which.

Interesting files are `actions/app-actions` and `constants/app-constants`, and the whole codebase can be grepped for `AppConstants\.` and `AppActions\.`. Pretty straight-forward but don't miss one!

**Comment by andynoelker** [comment]
I'd like to tackle this one as I want to familiarize myself with the code better and this is some low-hanging fruit. @fasterthanlime just to be clear, you want to make sure that all action names and constants are written so that the verb always comes first, followed by its object (i.e. `queue_cave`, `browse_game`, etc.)?

**Comment by fasterthanlime** [comment]
@andynoelker That's basically it!

The tricky part is, unfortunately, not every module that calls `AppActions.XXX` are covered by test cases - that's how url-store broke in v0.13.x for example ([here's the commit that fixes it](https://github.com/itchio/itch/commit/68c2c8c7dd0e14ed17a2b4232bfaac6c6a942634#diff-adafbaaf0c09a4aa4fd7108679d138f6L95))

It'd be possible to clean up action names just by carefully searching & replacing project-wide, but it would be 10x better to, at the same time, make sure all callsites are covered by tests - [like I did within the same release](https://github.com/itchio/itch/commit/528623bd9e05c760df2c57cd7bd339cf1b1d221f).

Overall coverage is tracked on our [coveralls.io page](https://coveralls.io/github/itchio/itch), but when developing locally, you can also run `npm run local-cov` and open `coverage/index.html` in a web browser.

Running all tests is annoying some times, so you can narrow it down with:

```
# run all specs in 'test/stores' - could also be a glob or a single file
# nyc can be installed system-wide with `npm install -g nyc`
# (might need to use sudo depending on your nodejs setup)
# `-s` option suppresses text output on nyc's first run, since we're
# telling it to generate HTML right after
nyc -s test/runner test/stores && nyc report -r html
```

TL;DR although simple, it can introduce annoying bugs & it's a good occasion to increase code coverage if you feel like it! I'd be happy to help / provide guidance for any of the above.

**Comment by andynoelker** [comment]
@fasterthanlime thanks for the clarification! While writing tests for all actions definitely makes this much less of a low-hanging fruit than I initially thought, it will be a fantastic opportunity for me to really understand this code better!

At my job, I often use React and Redux (which I like waaay better than vanilla Flux, no offense), but I'm pretty unfamiliar with a lot of the other technologies at use here like mori, nedb, tape, etc. So I will do my best to write good tests for every action, but expect a lot of noobish code in my first PR for this :)

**Comment by fasterthanlime** [comment]
> Not so low-hanging anmore

@andynoelker Feel free to submit a PR with all actions renamed & I'd simply merge it in a branch and add the tests myself :)

> Redux (which I like waaay better than vanilla Flux, no offense)

None taken - on the renderer side (libchromiumcontent), itch kinda shares Redux's philosophy: `app-store` manages everything, there's a single state object for the whole interface. Other stores exist in the 'browser' side only (node.js), but I'm edging towards the idea that maybe this is Flux abuse and they shouldn't be stores in the first place.

**Comment by andynoelker** [comment]
I will make an honest attempt to write some tests along with the name changes. I really want to understand this codebase better and that seems like a great way. That being said, I suspect you're going to want to revise some things when I make my first PR for this, so I look forward to your feedback! I really hope to contribute a lot more to this project!

> itch kinda shares Redux's philosophy

That's good to know! I have tried doing a custom Flux implementation in the past as well as playing around with [Alt](http://alt.js.org/) some, but Redux really clicked with me and just seems so much more maintainable and clear. Granted, I don't have enough knowledge of the itch app yet to have a proper comparison and learn the pros of a more vanilla Flux system, but I suspect that I will always lean towards consolidating state rather than fragmenting it.

Anyway, I probably won't have anything ready to show for this until next week, but I'm excited to delve deeper into this all!

**Comment by fasterthanlime** [comment]
> I will always lean towards consolidating state rather than fragmenting it.

Right, same feeling here. It's true that, for itch, we're operating under a different set of constraints:

As an Electron app, it has two sides - browser side (weird name for "modified node.js runtime") and renderer side (libchromiumcontent). They're independent processes, with both sync and async IPC (inter-process communication) facilities.

Both process types are quite fast, but passing data between them isn't. There's one browser process throughout the application's lifetime, and one renderer process per BrowserWindow (itch usually has only on BrowserWindow, except when purchasing).

Since windows can be entirely closed, with the app only living as a background process / tray icon (to actually save RAM when the app is minimzed to tray, rather than just hiding the window but keeping all its resources).

Synchronous IPC blocks both processes during the entire transfer, which wreaks havoc in Chromium's renderer timings for example, so it's not even an option. Neither is [RPC](https://github.com/atom/electron/blob/master/docs/api/remote.md) which uses sync IPC under the hood.

I've tried these solutions & eliminated them:
- Single browser(node)-side store, subscribed-to/queried from renderer-side via RPC
  - (Not only do callbacks leak, performance is atrocious as processes block a lot of the time)
- One store on each side, browser(node)-side is the canonical version, renderer(chromium)-side receives a JSON version of the data via async IPC.
  - (The app spends basically all its time doing JSON serialization/deserialization, and as far as the renderer-side is concerned, the whole state has changed all the time, making React `shouldComponentUpdate` useless, => terrible performance)

The current solution is a compromise, with a few stores on the browser(node)-side, which emit `diff` events so that app-store can maintain its version of the state with minimal IPC. That explains why our [app dispatcher](https://github.com/itchio/itch/blob/master/appsrc/dispatcher/app-dispatcher.js) looks a bit off - it has to do async IPC sometimes!

The amount of browser(node)-side stores is probably too high / overkill / not the right abstraction / Flux abuse, but I feel like it can be addressed incrementally.

(Checking again, it seems that https://github.com/itchio/itch/blob/master/appsrc/stores/window-store.js actually hides the window instead of closing it, currently. :panda_face:)

**Comment by fasterthanlime** [comment]
I went ahead and added a few more tests / cleaned up the codebase some more to hopefully lead the way :)

https://github.com/itchio/itch/compare/400a4bf3e8464fd484e44bb5284ba827a40bcf9b...HEAD

**Comment by andynoelker** [comment]
Wow thank you for that incredibly thorough explanation of the structure! I have no Electron experience and that is really helpful to know - it might be worth adding some of that info to the development doc at some point. It's so easy for someone who has no experience with the codebase  to just blindly assume that what worked for other React projects must apply here, but this really puts the project needs into perspective. Once I have a much better grasp on how the code and app work I should be able to contribute a little more to structural discussions. Until that time, I'll be sticking to smaller, more bite-sized pieces of code :)

And way to go on the new tests and the volume of your recent commits! You're unstoppable! :muscle:
